File: C11/백준/Silver/15657. N과 M （8）/N과 M （8）.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int compare(const void *a, const void *b){
	int A = *(const int *)a;
	int B = *(const int *)b;
	return A - B;
}

void backtrack(int n, int m, int depth, int array[], int arr[], bool visited[]){
	if(depth == m){
		for(int i = 0; i < m; i++){
			printf("%d ", arr[i]);
		}
		printf("\n");
		return;
	}
	for(int i = 1; i <= n; i++){
		if(!visited[i]){
			for (int j = i - 1; j >= 1; j--) {
                visited[j] = true;
            }
			arr[depth] = array[i];
			backtrack(n, m, depth + 1, array, arr, visited);
			for (int j = i - 1; j >= 1; j--) {
                visited[j] = false;
            }
		}
	}
}

int main(void){
	
	int N, M;
	scanf("%d %d", &N, &M);
	
	int array[10] = {0};
	bool visited[10] = { false };
	for(int i = 1; i <= N; i++){
		scanf("%d", &array[i]);
	}
	
	qsort(array, N + 1, sizeof(int), compare);
	int arr[10];
	
	backtrack(N, M, 0, array, arr, visited);
	
	return 0;
}

File: Java/백준/Silver/15657. N과 M （8）/N과 M （8）.java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {

        Scanner s = new Scanner(System.in);

        int N = s.nextInt(); //N개의 자연수 중에서
        int M = s.nextInt(); //M개를 고른 수열

        int[] array = new int[N + 1];;
        for (int i = 1; i <= N; i++) {
            array[i] = s.nextInt();
        }

        Arrays.sort(array);

        boolean[] visited = new boolean[N + 1];
        int[] arr = new int[M];
        backtrack(N, M, 0, array, arr, visited);

    }

    public static void backtrack(int n, int m, int depth, int[] array, int[] arr, boolean[] visited) {
        if (depth == m) {
            for (int i = 0; i < m; i++) {
                System.out.print(arr[i] + " ");
            }
            System.out.println();
            return;
        }
        for (int i = 1; i <= n; i++) {
            if (!visited[i]) {
                for (int j = i - 1; j >= 1; j--){
                    visited[j] = true;
                }
                arr[depth] = array[i];
                backtrack(n, m, depth + 1, array, arr, visited);
                for (int j = i - 1; j >= 1; j--){
                    visited[j] = false;
                }
            }
        }
    }
}